hbase-server/src/main/java/org/apache/hadoop/hbase/master/procedure/SnapshotProcedure.java

```diff
@@ -0,0 +1,395 @@
+    try {
+      ModifyRegionUtils.editRegions(exec, regions, new ModifyRegionUtils.RegionEditTask() {
+        @Override
+        public void editRegion(final RegionInfo regionInfo) throws IOException {
+          snapshotManifest.addRegion(CommonFSUtils.getTableDir(rootDir, snapshotTable), regionInfo);
+        }
+      });
+    } finally {
+      exec.shutdown();
+    }
+    status.setStatus("Completed referencing offline regions of table: " + snapshotTable);
+  }
+
+  private void snapshotMobRegion() throws IOException {
+    RegionInfo mobRegionInfo = MobUtils.getMobRegionInfo(htd.getTableName());
+    snapshotManifest.addMobRegion(mobRegionInfo);
+    status.setStatus("Completed referencing HFiles for the mob region of table: " + snapshotTable);
+  }
+
+  @Override
+  protected void serializeStateData(ProcedureStateSerializer serializer) throws IOException {
+    super.serializeStateData(serializer);
+    serializer.serialize(SnapshotProcedureStateData
+      .newBuilder().setSnapshot(this.snapshot).build());
+  }
+
+  @Override
+  protected void deserializeStateData(ProcedureStateSerializer serializer) throws IOException {
+    super.deserializeStateData(serializer);
+    SnapshotProcedureStateData data = serializer.deserialize(SnapshotProcedureStateData.class);
+    this.snapshot = data.getSnapshot();
+  }
+
+  private Procedure<MasterProcedureEnv>[] createRemoteSnapshotProcedures(MasterProcedureEnv env) {
+    return env.getAssignmentManager().getTableRegions(snapshotTable, true)
+      .stream().filter(r -> RegionReplicaUtil.isDefaultReplica(r))
+      .map(r -> new SnapshotRegionProcedure(snapshot, r))
+      .toArray(SnapshotRegionProcedure[]::new);
+  }
+
+  // here we assign region snapshot manifest verify tasks to all region servers
+  // in cluster with the help of master load balancer.
+  private Procedure<MasterProcedureEnv>[] createRemoteVerifyProcedures(MasterProcedureEnv env)
+      throws IOException {
+    List<RegionInfo> regions = env
+      .getAssignmentManager().getTableRegions(snapshotTable, false);
+    List<ServerName> servers = env
+      .getMasterServices().getServerManager().getOnlineServersList();
+    return env.getMasterServices().getLoadBalancer()
+      .roundRobinAssignment(regions, servers).entrySet().stream()
+      .map(e -> new SnapshotVerifyProcedure(snapshot, e.getValue(), e.getKey(), regions.size()))
+      .toArray(SnapshotVerifyProcedure[]::new);
+  }
+
+  @Override
+  public void toStringClassDetails(StringBuilder builder) {
+    builder.append(getClass().getName())
+      .append(", id=").append(getProcId())
+      .append(", snapshot=").append(ClientSnapshotDescriptionUtils.toString(snapshot));
+  }
+
+  public SnapshotDescription getSnapshotDesc() {
+    return snapshot;
+  }
+
+  @Override
+  protected void afterReplay(MasterProcedureEnv env) {
+    try {
+      prepareSnapshot(env);
+    } catch (IOException e) {
+      LOG.error("Failed replaying {}, mark procedure as failed", this, e);
```

Comment:
OK, so the rollback here is very simple, just delete the snapshot directory, so we are safe to mark it as failure at any time, no PONR, good.